Publish with custom site name

FadhlanR opened the pull request:
This PR ensures that publishing a realm with a custom site name works as expected. In addition, it fixes the margin issue we found by removing the background color from each domain option to match the design. It also implements an error message to display when publishing fails.


https://github.com/user-attachments/assets/9db6cac1-1b20-4661-bc20-429178e5b87a


https://github.com/user-attachments/assets/2f6ee85a-2fb2-443c-a066-b5517037dee2

FadhlanR committed:
Publish with custom site name

FadhlanR committed:
Fix matrix tests

FadhlanR committed:
lint fix

backspace commented:
I was trying out your branch via the preview, I’d already claimed a subdomain and it correctly detected that, but when I unclaimed it and tried to claim it again, this happened:

<img width="3258" height="2174" alt="minicatalog 2025-10-23 09-39-31" src="https://github.com/user-attachments/assets/e4b4d137-046a-486d-90fe-6d58fda1f630" />

I’ll look into it further, I don’t think it relates to your PR here, just making note. Will proceed with review now!

backspace commented on packages/host/app/components/operator-mode/host-submode.gts:
```diff
@@ -43,14 +43,20 @@ export default class HostSubmode extends Component<HostSubmodeSignature> {
   @tracked isPublishingRealmPopoverOpen = false;
   @tracked isOpenSitePopoverOpen = false;
+  @tracked publishError: string | null = null;
+  @tracked publishErrors: Map<string, string> = new Map();
```
Is it possible to combine `publishError` into `publishErrors` somehow? It seems clumsy to me to have two of these passed around everywhere and reset.

backspace commented on packages/host/app/components/operator-mode/host-submode.gts:
```diff
@@ -112,7 +118,27 @@ export default class HostSubmode extends Component<HostSubmodeSignature> {
+
+      // Trigger reactivity by creating a new Map instance
+      this.publishErrors = new Map(this.publishErrors);
+    }
```
Since this is an Ember Concurrency task, it seems possible to encapsulate the errors into the task’s `last` run, as described [here](https://ember-concurrency.com/docs/derived-state). Then the error state can be detected with `handlePublish.last.isError` and the error(s) accessed with `handlePublish.last.error`. This seems more elegant to me.

backspace commented on packages/host/app/components/operator-mode/publish-realm-modal.gts:
```diff
@@ -468,6 +481,13 @@ export default class PublishRealmModal extends Component<Signature> {
+    if (this.claimedDomainPublishedUrl) {
+      window.open(this.claimedDomainPublishedUrl + 'index', '_blank');
+    }
+  }
```
I failed to notice this before but I’d prefer these buttons be true links vs using `window.open`. That lets you hover over the button and see the URL. I don’t see any value in doing this via redirect. (Same for the above.)

backspace commented on packages/host/app/components/operator-mode/publish-realm-modal.gts:
Let’s rename this now that it’s not the only site?

backspace commented on packages/host/app/components/operator-mode/publish-realm-modal.gts:
```diff
@@ -712,19 +820,43 @@ export default class PublishRealmModal extends Component<Signature> {
+                <IconX width='12' height='12' />
+              </BoxelButton>
+            </div>
+          {{/if}}
```
How does this get rendered? I didn’t see it in your videos. It doesn’t seem to be covered by a test.

FadhlanR replied:
I removed this part, this is not needed.

backspace commented on packages/host/tests/acceptance/host-submode-test.gts:
```diff
@@ -1012,6 +1012,298 @@ module('Acceptance | host submode', function (hooks) {
+            );
+            assert.strictEqual(target, '_blank', 'Open in a new tab');
+            return null;
+          };
```
Another benefit of using true links (styled as buttons) is that you don’t need to mock `window.open`, you can just check the value of `a.href`.

backspace reviewed (changes requested):
I tried this out and it worked great, I have some suggestions for implementation improvement though.

I’m not sure if this is related to your PR but notice the alignment of the Claim button here, could it be grid-related?

<img width="721" height="534" alt="Image" src="https://github.com/user-attachments/assets/37ac53e4-3447-44a0-96e8-65346a6abc3c" />

It’s meant to be aligned with the field:

<img width="481" height="491" alt="Image" src="https://github.com/user-attachments/assets/43e253d0-f86a-40e1-9f6a-8d2509fe4502" />

backspace commented:
I do think the button alignment is a problem with this PR because this is how it looks on staging:

<img width="2730" height="2110" alt="oct2 2025-10-23 11-01-45" src="https://github.com/user-attachments/assets/e1efaba9-6683-467f-bfb5-9fb37d1a4902" />

(The state appearing valid despite the field being empty is a bug though, I’ll open an issue.)

FadhlanR committed:
Fix claim site button position

FadhlanR committed:
Refactor error handling

FadhlanR committed:
Update open site using links

FadhlanR commented:
> I do think the button alignment is a problem with this PR because this is how it looks on staging:
> 
> <img alt="oct2 2025-10-23 11-01-45" width="2000" height="2000" src="https://private-user-images.githubusercontent.com/43280/504665157-e1efaba9-6683-467f-bfb5-9fb37d1a4902.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEyMzAwNzIsIm5iZiI6MTc2MTIyOTc3MiwicGF0aCI6Ii80MzI4MC81MDQ2NjUxNTctZTFlZmFiYTktNjY4My00NjdmLWJmYjUtOWZiMzdkMWE0OTAyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIzVDE0MjkzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0ODZhMTg3ODRlNzY2MTEwZjhhODBmZGEzNmZkMmI5MzZiZGI5YTVhMWE5MTMyMWY0MzJhYTMyNjY3NGEyNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MmLkknNLEiYrpKkNKEWKrNGU_0oMjjYcp5N75zRUZwY">
> (The state appearing valid despite the field being empty is a bug though, I’ll open an issue.)

Really appreciate your feedback. I fixed the position of the “Claim Site” button, refactored the error handling using `task.last.error` so it looks cleaner, and changed the “Open Site” button to be a link.

backspace commented on packages/host/app/components/operator-mode/host-submode.gts:
```diff
@@ -111,32 +112,78 @@ export default class HostSubmode extends Component<HostSubmodeSignature> {
+    // TODO: remove this after this issue (https://github.com/machty/ember-concurrency/issues/40)
+    // is resolved.
+    promise.catch((_error) => {});
+  }
```
I forgot about or didn’t know about this, unpleasant 😞 but I think the error encapsulation is still worth it

backspace reviewed (approved):
thanks for the updates, I tried it again via the preview, excellent 🎉

habdelra commented:
@codex review
